pkg/apis/core/validation/validation.go

```diff
@@ -1867,6 +1867,13 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume) field.E
 	}
 	allErrs = append(allErrs, ValidateImmutableField(newPv.Spec.VolumeMode, oldPv.Spec.VolumeMode, field.NewPath("volumeMode"))...)
 
+	oldSize := oldPv.Spec.Capacity["storage"]
+	newSize := newPv.Spec.Capacity["storage"]
+
+	if newSize.Cmp(oldSize) < 0 {
```

Comment:
I'm not sure if we can actually tighten validation, even though it's technically a bug fix: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#backward-compatibility-gotchas

Unfortunately it seems like with validation, we must get it right the first time :(  

The best we can maybe do here is to have the controller return a better error message?  Although @gnufied, does this have any implications on quota calculations?

Reply:
@gnufied do you have any other suggestions?

Reply:
> what are the consequences if we allow this to shrink?
> 
> > does this have any implications on quota calculations?
> 
> quota only applies to namespaced resources, not PVs

pvc request <= pv capacity when binding
if pv shrinks and pv capacity < pvc request , pvc should release this pv or not? Or pvc should also shrink its request ?

Reply:
> if pv shrinks and pv capacity < pvc request , pvc should release this pv or not? Or pvc should also shrink its request ?

I would not expect the PVC to change or be unbound.

Does the mismatch in size between the PVC spec and the PV trigger a resize request?

Reply:
Yes, if [PV size is less than PVC](https://github.com/kubernetes/kubernetes/blob/a381f7cb3efa64395661c73961c07b02594acf64/pkg/volume/util/operationexecutor/operation_generator.go#L1519), then resize will be triggered. Operations are supposed to be idempotent, so that shouldn't be a huge deal.

Reply:
hrmm. would that cause problems for PVs whose drivers don't support resize?

Reply:
It seems we only support expansion now. shrink would be forbidden when controller and volume manager check the capacity.

Reply:
> It seems we only support expansion now. shrink would be forbidden when controller and volume manager check the capacity.

Would they have the info to know the PV shrank, or would it appear as if the PV should grow? Do they only have the PVC and PV info, or do they actually have underlying info that tells them the volume is bigger than recorded in the PV?

Reply:
cc @gnufied

Reply:
The mismatch between PV & PVC size in itself will not trigger resize operation. The code won't even enter operation_generator without - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/expand/expand_controller.go#L176 

Same check exists in external resize-controller too.

Reply:
In general, write access to PV spec is highly privileged. We don't allow changing type (volumesource is immutable), but most other things are. If this doesn't cause critical issues (like data loss, etc), I'm inclined to leave these as-is.

Reply:
If so, I would like close this PR